repository/src/lib.rs

```diff
@@ -168,9 +168,206 @@ impl<T: RawRepository> DistributedRepository<T> {
     pub async fn fetch(
         &mut self,
         _network_config: &NetworkConfig,
-        _known_peers: &[Peer],
+        known_peers: &[Peer],
     ) -> Result<(), Error> {
-        unimplemented!()
+        // Add remote repo and check duplicated ones
+        for _peer in known_peers {
+            let existing_remote = self.raw.list_remotes().await?;
+            let remote_name = "yoonho".to_string();
+            let remote_url = "github".to_string();
+            if !existing_remote
+                .iter()
+                .map(|(remote_name, _)| remote_name)
+                .any(|x| x == &remote_name)
+            {
+                self.raw.add_remote(remote_name, remote_url).await?;
+            }
+            // TODO: change yoonho, github into something else with peer
+            // It will be evaluated by "git" method
+        }
+
+        self.raw.fetch_all().await?;
+
+        // For all incoming branches, verify all the incoming commits and apply them to the local repository only if they are valid.
+        let branches = self.raw.list_remote_tracking_branches().await?;
+        let branches: Vec<&String> = branches.iter().map(|(branch, _, _)| branch).collect();
+        let finalized_branch_commit_hash =
+            self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+        let last_header = self.get_last_finalized_block_header().await?;
+        let reserved_state = self.get_reserved_state().await?;
+
+        'branch: for branch in branches {
+            let branch_tip_commit_hash = self.raw.locate_branch(branch.clone()).await?;
+            // Check if the branch is rebased on top of the `finalized` branch.
+            if finalized_branch_commit_hash
+                != self
+                    .raw
+                    .find_merge_base(branch_tip_commit_hash, finalized_branch_commit_hash)
+                    .await?
+            {
+                self.raw.delete_branch(branch.to_string()).await?;
+                continue 'branch;
+            }
+
+            // Construct a commit list starting from the next commit of the last finalized block to the `branch_tip_commit`(the most recent commit of the branch)
```

Comment:
Please update this process using `query_commit_path`.

refer to #197

Reply:
Also trim this long comment please

refer to https://github.com/postech-dao/simperby/commit/f65178195edfca265f3db0ab2ae613fc746b7df9

Reply:
Fixed.